var/spack/repos/builtin/packages/mfem/package.py

```diff
@@ -115,14 +131,23 @@ def install(self, spec, prefix):
                 'METIS_OPT=-I%s' % spec['metis'].prefix.include,
                 'METIS_LIB=%s' % metis_lib])
 
-        if '+mpi' in spec:
+        if 'mpi' in spec:
```

Comment:
This actually does work because it checks the presence of a dependency, not the value of the variant.  `if 'mpi' in spec` will be `True` if any `mpi` provider is a dependency of this package.  Since `mfem`` depends_on('mpi', when='+mpi'), it will depend on an MPI implementation when it's `+mpi`, and it won't when it is `~mpi`.  Same for the other two, as well.